flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapReducingState.java

```diff
@@ -68,62 +63,29 @@ public HeapReducingState(
 
 	@Override
 	public V get() {
-		Preconditions.checkState(currentNamespace != null, "No namespace set.");
-		Preconditions.checkState(backend.getCurrentKey() != null, "No key set.");
-
-		Map<N, Map<K, V>> namespaceMap =
-				stateTable.get(backend.getCurrentKeyGroupIndex());
-
-		if (namespaceMap == null) {
-			return null;
-		}
-
-		Map<K, V> keyedMap = namespaceMap.get(currentNamespace);
-
-		if (keyedMap == null) {
-			return null;
-		}
-
-		return keyedMap.get(backend.<K>getCurrentKey());
+		return stateTable.get(currentNamespace);
 	}
 
 	@Override
 	public void add(V value) throws IOException {
-		Preconditions.checkState(currentNamespace != null, "No namespace set.");
-		Preconditions.checkState(backend.getCurrentKey() != null, "No key set.");
+		final N namespace = currentNamespace;
 
 		if (value == null) {
 			clear();
 			return;
 		}
 
-		Map<N, Map<K, V>> namespaceMap =
-				stateTable.get(backend.getCurrentKeyGroupIndex());
-
-		if (namespaceMap == null) {
-			namespaceMap = createNewMap();
-			stateTable.set(backend.getCurrentKeyGroupIndex(), namespaceMap);
-		}
-
-		Map<K, V> keyedMap = namespaceMap.get(currentNamespace);
-
-		if (keyedMap == null) {
-			keyedMap = createNewMap();
-			namespaceMap.put(currentNamespace, keyedMap);
-		}
-
-		V currentValue = keyedMap.put(backend.<K>getCurrentKey(), value);
+		final StateTable<K, N, V> map = stateTable;
+		final V currentValue = map.putAndGetOld(namespace, value);
```

Comment:
Why not push the Reduce into the table? `map.reduce(namespace, value, ReduceFunction)`? That gives the minimal number of navigations possible, similar to using `putIfAbsent()` rather then a `get()/put()`